text/0014-configuration.md

```diff
@@ -0,0 +1,78 @@
+structured, what keys it must have, and constrains what types of values are allowed.
+
+A **config library** is a code module that allows developers to define config 
+schemas, and loads config files in accordance with that schema. It may also 
+provide other tooling related to configuration.
+
+
+## Decision
+
+Microfrontend developers should use the official OpenMRS MF Config Library. 
+The Config Library should be a JavaScript module that
+*   Provides an API for modules to define a config schema
+*   Provides an API for modules to access config values
+*   Looks for config files in an implementation’s esm-root-config (or using the Import Map)
+*   Validates config files in non-production environments
+*   Generates documentation for config schemas
+*   Provides an API for an implementation’s esm-root-config to specify a config file or a hierarchy of config files
+*   Provides a UI for viewing the resultant configuration, and which configuration values come from which files, and providing temporary overrides (similar to how Single-SPA allows Import Map overrides)
+
+The config library should require that each configuration element has a default value.
+
+The specifics of what these APIs should look like are the topic of a [separate RFC](https://docs.google.com/document/d/1Srazq1xfZSmIE7TfyMNKAriPgV2olEMJ68CirYpbgME/edit).
+
+
+## Reason for Decision
+
+For having a config library
+*   **Configuration as a first-class citizen**. We should do everything we can to encourage developers to make their modules configurable rather than producing implementation-specific modules. Providing a standard mechanism to do so will help.
+*   **Good error messages for bad configs**. A config library allows defining constraints on the acceptable config values. This allows the application to fail fast with good error messages. Good error messages are essential because we expect non-dev implementers to be doing configuration. We should not expect configurers to be able to debug code.
+*   **Generate config documentation**. It’s hard to keep docs complete and up to date. It would be much easier if nicely-formatted documentation could be generated from the schema.
+*   **Enforce conventions**. The config library can enforce certain conventions about, for example, config namespacing or acceptable ways to reference concepts.
+*   **Validation during config-writing, speed in production**. The config library can execute complex and potentially expensive validation while the implementer is working on the configuration. In production, validation can be turned off to reduce load time.
+*   **Possible future support for other config file types**. While SystemJS provides native support for JSON import, it would be great to allow implementers to write config files that are less verbose, easier to work with, and support comments. Good candidates include HCL and YAML.
+
+For having defaults for all values
+*   Eliminate boilerplate.
+
+For having one combined config file
```

Comment:
I am thinking of pluggability, in line with smart-on-fhir apps. Imagine an openmrs app store (e.g. Odoo apps as well) wherefrom I can just add an app, and that app gets downloaded and plugged in. In that case, it would be the case, that the app would bring its own config, which would probably come in their esm module and not requiring any npm build. We should keep our design open. Also reason for also supporting separate config esms for modules.